CI: ci-integration: TestLocalObserverServer_NodeLabels

## CI failure

```
--- FAIL: TestLocalObserverServer_NodeLabels (0.00s)

    local_observer_test.go:699: 
        	Error Trace:	/home/runner/work/cilium/cilium/pkg/hubble/observer/local_observer_test.go:699
        	            				/home/runner/work/cilium/cilium/pkg/hubble/testutils/fake.go:43
        	            				/home/runner/work/cilium/cilium/pkg/hubble/observer/local_observer.go:369
        	            				/home/runner/work/cilium/cilium/pkg/hubble/observer/local_observer_test.go:737
        	Error:      	Not equal: 
        	            	expected: []string{"kubernetes.io/arch=amd64", "kubernetes.io/hostname=ip-1-2-3-4.us-west-2.compute.internal", "kubernetes.io/os=linux", "topology.kubernetes.io/region=us-west-2", "topology.kubernetes.io/zone=us-west-2d"}
        	            	actual  : []string(nil)
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,8 +1,2 @@
        	            	-([]string) (len=5) {
        	            	- (string) (len=24) "kubernetes.io/arch=amd64",
        	            	- (string) (len=60) "kubernetes.io/hostname=ip-1-2-3-4.us-west-2.compute.internal",
        	            	- (string) (len=22) "kubernetes.io/os=linux",
        	            	- (string) (len=39) "topology.kubernetes.io/region=us-west-2",
        	            	- (string) (len=38) "topology.kubernetes.io/zone=us-west-2d"
        	            	-}
        	            	+([]string) <nil>
        	            	 
        	Test:       	TestLocalObserverServer_NodeLabels
```

- Observed in https://github.com/cilium/cilium/pull/33237
- CI run: https://github.com/cilium/cilium/actions/runs/9594579558/job/26457473354
- Likely cause https://github.com/cilium/cilium/pull/32851

--- a/daemon/cmd/hubble.go
+++ b/daemon/cmd/hubble.go
@@ -148,7 +148,11 @@ func (d *Daemon) launchHubble() {
 
 	// fill in the local node information after the dropEventEmitter logique,
 	// but before anything else (e.g. metrics).
-	localNodeWatcher := observer.NewLocalNodeWatcher(d.ctx, d.nodeLocalStore)
+	localNodeWatcher, err := observer.NewLocalNodeWatcher(d.ctx, d.nodeLocalStore)
+	if err != nil {
+		logger.WithError(err).Error("Failed to retrieve local node information")
+		return
+	}
 	observerOpts = append(observerOpts, observeroption.WithOnDecodedFlow(localNodeWatcher))
 
 	grpcMetrics := grpc_prometheus.NewServerMetrics()

--- a/pkg/hubble/observer/local_node_watcher.go
+++ b/pkg/hubble/observer/local_node_watcher.go
@@ -7,8 +7,6 @@ import (
 	"context"
 	"slices"
 
-	"github.com/cilium/stream"
-
 	flowpb "github.com/cilium/cilium/api/v1/flow"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/node"
@@ -17,9 +15,8 @@ import (
 // LocalNodeWatcher populate Hubble flows local node related fields (currently
 // only labels).
 type LocalNodeWatcher struct {
-	mu      lock.Mutex
-	updated int // update counter
-	cache   struct {
+	mu    lock.Mutex
+	cache struct {
 		// labels are represented as a Go map in node.LocalNode, but we need a
 		// key=val slice for Hubble flows.
 		labels []string
@@ -27,21 +24,29 @@ type LocalNodeWatcher struct {
 }
 
 // NewLocalNodeWatcher return a new LocalNodeWatcher. The given context control
-// whether the LocalNodeWatcher gets updated by the localNodeStream. It is safe
+// whether the LocalNodeWatcher gets updated by the localNodeStore. It is safe
 // to use the returned LocalNodeWatcher once the context is cancelled, but its
 // information might be out-of-date.
-func NewLocalNodeWatcher(ctx context.Context, localNodeStream stream.Observable[node.LocalNode]) *LocalNodeWatcher {
+func NewLocalNodeWatcher(ctx context.Context, localNodeStore *node.LocalNodeStore) (*LocalNodeWatcher, error) {
 	watcher := LocalNodeWatcher{}
-	localNodeStream.Observe(ctx, watcher.update, watcher.complete)
-	return &watcher
+	// fetch the initial local node
+	n, err := localNodeStore.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	watcher.update(n)
+	// subscribe to local node changes.
+	localNodeStore.Observe(ctx, watcher.update, watcher.complete)
+
+	return &watcher, nil
 }
 
 // OnDecodedFlow implements OnDecodedFlow for LocalNodeWatcher. The
 // LocalNodeWatcher populate the flow's node_labels field.
 func (w *LocalNodeWatcher) OnDecodedFlow(_ context.Context, flow *flowpb.Flow) (bool, error) {
 	w.mu.Lock()
+	defer w.mu.Unlock()
 	flow.NodeLabels = w.cache.labels
-	w.mu.Unlock()
 	return false, nil
 }
 
@@ -49,24 +54,15 @@ func (w *LocalNodeWatcher) OnDecodedFlow(_ context.Context, flow *flowpb.Flow) (
 func (w *LocalNodeWatcher) update(n node.LocalNode) {
 	labels := sortedLabelSlice(n.Labels)
 	w.mu.Lock()
-	w.cache.labels = labels
-	w.updated++
-	w.mu.Unlock()
-}
-
-// generation return this watcher update counter.
-func (w *LocalNodeWatcher) generation() int {
-	w.mu.Lock()
 	defer w.mu.Unlock()
-	return w.updated
+	w.cache.labels = labels
 }
 
 // complete clears the LocalNodeWatcher cache.
 func (w *LocalNodeWatcher) complete(error) {
 	w.mu.Lock()
+	defer w.mu.Unlock()
 	w.cache.labels = nil
-	w.updated++
-	w.mu.Unlock()
 }
 
 // sortedLabelSlice convert a given map of key/val labels, and return a sorted

--- a/pkg/hubble/observer/local_node_watcher_test.go
+++ b/pkg/hubble/observer/local_node_watcher_test.go
@@ -6,7 +6,6 @@ package observer
 import (
 	"context"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/require"
 
@@ -57,11 +56,10 @@ func Test_LocalNodeWatcher(t *testing.T) {
 	var watcher *LocalNodeWatcher
 
 	t.Run("NewLocalNodeWatcher", func(t *testing.T) {
-		watcher = NewLocalNodeWatcher(ctx, node.NewTestLocalNodeStore(localNode))
+		var err error
+		watcher, err = NewLocalNodeWatcher(ctx, node.NewTestLocalNodeStore(localNode))
+		require.NoError(t, err)
 		require.NotNil(t, watcher)
-		require.Eventually(t, func() bool {
-			return watcher.generation() > 0
-		}, time.Second, 10*time.Millisecond)
 	})
 
 	t.Run("OnDecodedFlow", func(t *testing.T) {

--- a/pkg/hubble/observer/local_observer_test.go
+++ b/pkg/hubble/observer/local_observer_test.go
@@ -689,14 +689,22 @@ func TestLocalObserverServer_NodeLabels(t *testing.T) {
 			},
 		},
 	}
-	localNodeWatcher := NewLocalNodeWatcher(ctx, node.NewTestLocalNodeStore(localNode))
+	localNodeWatcher, err := NewLocalNodeWatcher(ctx, node.NewTestLocalNodeStore(localNode))
+	require.NoError(t, err)
+	require.NotNil(t, localNodeWatcher)
 
 	// fake hubble server setup.
 	flowsReceived := 0
 	req := &observerpb.GetFlowsRequest{Number: uint64(1)}
 	fakeServer := &testutils.FakeGetFlowsServer{
 		OnSend: func(response *observerpb.GetFlowsResponse) error {
-			assert.Equal(t, localNodeWatcher.cache.labels, response.GetFlow().GetNodeLabels())
+			// NOTE: a bit hacky to directly access the localNodeWatcher cache,
+			// but we don't have any use yet for an accessor method beyond this
+			// package local test.
+			localNodeWatcher.mu.Lock()
+			expected := localNodeWatcher.cache.labels
+			localNodeWatcher.mu.Unlock()
+			assert.Equal(t, expected, response.GetFlow().GetNodeLabels())
 			flowsReceived++
 			return nil
 		},

